### ISSUE TITLE
Request for Access Token return 400 Bad Request on Android

### ISSUE THREAD

**Issue by ctl018** [alexa/alexa-avs-sample-app#395]
I'm trying to get a prototype Android AVS app working by cobbling together this sample Alexa source code. After trying for more than a week, I still cannot get the authentication to work. I am running from the Android emulator.

Basically I try to get an Authorization Code Grant using Android app to do what was implemented via CompanionApp+JavaClient in the Alexa sample app.
In Alexa sample app:
1) AndroidCompanionApp --> request device info --> JavaClient
2) AndroidCompanionApp --> request authorization code (via LWA SDK) --> avs-alexa-na.amazon.com
3) AndroidCompanionApp --> send AuthorizeResult data --> JavaClient
4) JavaClient --> request access token (HTTP POST) --> api.amazon.com/auth/o2/token
At the end of step 4, JavaClient gets an Access Token.

In my Android AVS app:
1) Device ID and code challenge is done in app (this step is skipped)
2) Android AVS app --> request authorization code (via LWA SDK) --> avx-alexa-na.amazon.com
3) OAuth2ClientForPkce.java is copied from JavaClient and used in Android app (this step is skipped)
4) Android AVS app --> request access token (HTTP POST) --> api.amazon.com/auth/o2/token
In step 4, I get HTTP code 400 Bad Request result.

Below is the log output of the error that I get.
Avs_test: AuthorizeListener.onSuccess: uri amzn://com.example.peter.test_avs
Avs_test: AuthorizeListener.onSuccess: authcode ANeNftJMCJLcqYZjLJcu
Avs_test: AuthorizeListener.onSuccess: clientid amzn1.application-oa2-client.xxxxxx
Avs_test: cliend ID = amzn1.application-oa2-client.xxxxxx
Avs_test: Auth code = ANeNftJMCJLcqYZjLJcu
Avs_test: redirect URI = amzn://com.example.peter.test_avs
OAuth2ClientForPkce: exchangeAuthCodeForTokens json {"grant_type":"authorization_code","code":"ANeNftJMCJLcqYZjLJcu","redirect_uri":"amzn://com.example.peter.test_avs","client_id":"amzn1.application-oa2-client.xxxxxx","code_verifier":"yyyyyy"}
OAuth2ClientForPkce: exchangeAuthCodeForTokens con https://api.amazon.com/auth/o2/token
OAuth2ClientForPkce: postRequest https://api.amazon.com/auth/o2/token
OAuth2ClientForPkce: com.android.okhttp.internal.huc.HttpURLConnectionImpl:https://api.amazon.com/auth/o2/token
OAuth2ClientForPkce: {"grant_type":"authorization_code","code":"ANeNftJMCJLcqYZjLJcu","redirect_uri":"amzn://com.example.peter.test_avs","client_id":"amzn1.application-oa2-client.xxxxxx","code_verifier":"yyyyyy"}
OAuth2ClientForPkce: connect error 400 = Bad Request
                     buffer(com.android.okhttp.okio.GzipSource@ab14e21).inputStream()

Can anyone point why is the amazon server returning bad request in my app but has no problem when same thing is done in JavaClient?

My device security profile as set the web-settings according the instructions in Alexa developer site:
https://localhost:3000
https://localhost:3000/authresponse

**Comment by keleric** [comment]
Please see: https://forums.developer.amazon.com/articles/10110/creating-an-android-avs-app-vs-creating-an-android.html

You'll want to follow the steps to create an AVS app.

**Comment by ctl018** [comment]
Yes, I have read that article and also tried that example. 
There are 2 experiences that I had when trying the implicit grant method which steered me towards the authorization grant method:
1. Occasionally, after clicking the "Continue" button in the Sign In process, I get a page which says:
       Login with amazon
       We're sorry!
       An error occurred when we tried to process your request. Rest assured, we're already
       working on the problem and expect to resolve it shortly
2. The user will need to sign in every time whenever the implicit grant Access Token expires or when the app is killed or when the android mobile device is rebooted (in my case the emulator had to be restarted because it hung up).

So my original question is why would sending an Access Token request via HTTP POST (via Android) result in an error 400 Bad Request response from the LWA server when the same code to do it works in the sample JavaClient?

**Comment by ctl018** [comment]
Ok, finally managed to get the Access Tokens using the Authorization Grant method running entirely in the Android app.

Turns out the Bad Request HTTP POST response was due to some problem with the Code Verifier string generated during in the call to CodeChallengeWorkflow.generateProofKeyParameters().

In Android, I was using android.util.Base64 instead of org.apache.commons.codec.binary.Base64 that was in the original JavaClient code. To get it to work, I swapped with com.google.api.client.util.Base64 instead.
